codegen/smithy-go-codegen/src/main/java/software/amazon/smithy/go/codegen/UnionGenerator.java

```diff
@@ -51,48 +52,41 @@ public void run() {
-                writer.write("Value() $T", memberSymbol);
+            writer.openBlock("type $L struct {", "}", exportedMemberName, () -> {
+                // Union members can't have null values, so for simple shapes we don't
+                // use pointers. We have to use pointers for complex shapes since,
+                // for example, we could still have a map that's empty or which has
+                // null values.
+                if (target instanceof SimpleShape) {
+                    writer.write("Value $T", memberSymbol);
+                } else {
+                    writer.write("Value $P", memberSymbol);
+                }
             });
 
-            writer.openBlock("func As$L(v $T) $L {", "}", exportedMemberName, memberSymbol, exportedMemberName, () -> {
-                writer.write("return $L{value: v}", unExportedMemberName);
-            });
-
-            // Create the member's un-exported concrete type
-            writer.openBlock("type $L struct {", "}", unExportedMemberName, () -> {
-                writer.write("value $T", memberSymbol);
-            });
-
-            writer.write("func ($L) is$L() {}", unExportedMemberName, symbol.getName());
-            writer.write("func ($L) is$L() {}", unExportedMemberName, exportedMemberName);
-            writer.openBlock("func (v $L) Value() $T {", "}", unExportedMemberName, memberSymbol, () -> {
-                writer.write("return v.value");
-            });
+            writer.write("func ($L) is$L() {}", exportedMemberName, symbol.getName());
         }
 
         // Creates a fallback type for use when an unknown member is found. This
         // could be the result of an outdated client, for example.
-        writer.writeDocs(symbol.getName()
-                + "Unknown is returned when a union member is returned over the wire, but has an unknown tag.");
-        writer.openBlock("type $LUnknown struct {", "}", symbol.getName(), () -> {
-            // The tag (member) name received over the wire. Necessary for re-serialization.
-            writer.write("tag string");
+        String unknownStructName = symbol.getName() + "Unknown";
+        writer.writeDocs(unknownStructName
+                + " is returned when a union member is returned over the wire, but has an unknown tag.");
+        writer.openBlock("type $L struct {", "}", unknownStructName, () -> {
+            // The tag (member) name received over the wire.
+            writer.write("Tag string");
             // The value received.
-            writer.write("value []byte");
+            writer.write("Value []byte");
         });
-        writer.write("func (v $LUnknown) is$L() {}", symbol.getName(), symbol.getName())
-                .write("func (v $LUnknown) Value() []byte { return v.value }", symbol.getName())
-                .write("func (v $LUnknown) Tag() string { return v.tag }", symbol.getName());
+
+        writer.write("func (v $L) is$L() {}", unknownStructName, symbol.getName());
```

Comment:
Is it possible to only generate a single one of these unknown member structs, instead of one for ever union defined?  The definition between them will not change. Each API client's type package wopuld need its own `UnknownUnionMember` struct generated, with a method for each union type in the model added.

```go
// UnknownUnionMember
type UnknownUnionMember struct {
    Tag string
    Value []byte
}
```

For each union in an  API a `is` method would be generated, e.g. for unions `Foo`, `Baz`, `Bar`.

```go
func (UnknownUnionMember) isFoo() {}
func (UnknownUnionMember) isBaz() {}
func (UnknownUnionMember) isBar() {}
```